### PR TITLE
drivers: dma: Add initial support for NXP SDMA

### PR DESCRIPTION
**Pull request by dbaluta**
This adds initial support for NXP SDMA controller. We make use
of NXP HAL to configure the IP.    SDMA uses BD (buffer descriptors) to describe a transfer. We create
a cyclic list of descriptors and trigger them manually at start and later when data is available.
This is tested using Sound Open Firmware app on top of Zephyr.

We first need to merge following PRs:

- [x] https://github.com/zephyrproject-rtos/zephyr/pull/80327
- [x] https://github.com/zephyrproject-rtos/hal_nxp/pull/458
- [x] https://github.com/zephyrproject-rtos/hal_nxp/pull/394

### CODE REVIEW

**Comment by dbaluta** [comment]
Changes since v1:
- addressed @iuliana-prodan  comments

File: drivers/dma/dma_nxp_sdma.c [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+	if (chan_id >= FSL_FEATURE_SDMA_MODULE_CHANNEL)
+		return false;
+
+	dev_data->chan[chan_id].event_source = *((int*)param);
+	dev_data->chan[chan_id].index = chan_id;
+
+	return true;
+}
+
+static const struct dma_driver_api sdma_api = {
```
Comment (pdgendt) [inline drivers/dma/dma_nxp_sdma.c:414]: All device drivers need to use the `DEVICE_API` macro.

```suggestion
static DEVICE_API(dma, sdma_api) = {
```

Comment (dbaluta) [reply drivers/dma/dma_nxp_sdma.c:414]: thanks will fix.

**Comment by dbaluta** [comment]
Changes since v2:
- use DEVICE_API as per @pdgendt  comment.

**Review by LaurentiuM1234 — DISMISSED** [review]

**Review by iuliana-prodan — DISMISSED** [review]

File: drivers/dma/dma_nxp_sdma.c [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+		*type = config->dma_slot;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+void dma_nxp_sdma_callback(sdma_handle_t *handle, void *userData, bool TransferDone,
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply drivers/dma/dma_nxp_sdma.c:175]: When a block is completed the SDMA IP will raise an interrupt.  We handle the interrupt using `dma_nxp_sdma_isr` (we use the NXP HAL API to clear / handle the interrupt):
```
 -> dma_nxp_sdma_isr
     -> SDMA_ClearChannelInterruptStatus 
     -> SDMA_HandleIRQ (function from NXP HAL)
         -> dma_nxp_sdma_callback
```         
Then the dma_nxp_sdma_callback prepares the next BD to be transferred.

Is there any Zephyr DMA API specific thing that I'm missing?

File: drivers/dma/dma_nxp_sdma.c [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+			config->source_data_size, block_cfg->block_size,
+			is_last, true, is_wrap, chan_data->transfer_cfg.type);
+
+		chan_data->capacity += block_cfg->block_size;
+		block_cfg = block_cfg->next_block;
+		crt_bd++;
+	}
+}
+
+static int dma_nxp_sdma_config(const struct device *dev, uint32_t channel,
```
[the remark this answers is not in the document]
Comment (dbaluta) [reply drivers/dma/dma_nxp_sdma.c:259]: > This should support the callback to allow a user to understand when a block or sequence of transfers is done (though this seems to be a one shot dma is that right?)

Yes, it is useful to think as DMA operating in one shot mode to easier understand how it works.

The flow goes like this:

- At init we register `dma_nxp_sdma_isr` to handle a transfer done event.
- `sdma_nxp_sdma_config`  configures `DMA_NXP_SDMA_BD_COUNT` buffer descriptors to be used for transfer (by default this is `2`).
- `dma_nxp_sdma_start` manually starts transfer of the first buffer descriptor
- Once the first BD is completed `dma_nxp_sdma_isr` is called to submit the next BD to be send. This goes over and over preparing the next BD until the transfer is stopped or suspended.

File: drivers/dma/dma_nxp_sdma.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (teburd) [reply drivers/dma/dma_nxp_sdma.c:175]: yes, the config lets a user specified callback to attach to this exact event (either block or full transfer complete event). How does this DMA ever notify the DMA user that a transfer is complete? It doesn't seem to do this in any possible way other than polling the DMA status?

File: drivers/dma/dma_nxp_sdma.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (teburd) [reply drivers/dma/dma_nxp_sdma.c:259]: That's fine, but how do you know when a block is done here from the DMA user perspective? The only method I see for this driver is polling dma status which isn't the expected design around Zephyr's DMA API afaik, we have the ability to attach a callback to either individual block completions or a transfer completion (sequence of blocks).

Polling of course is also possible, but this driver doesn't allow for attaching a callback to those events as it should.

File: drivers/dma/dma_nxp_sdma.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (dbaluta) [reply drivers/dma/dma_nxp_sdma.c:175]: @teburd  exactly. 

SOF design uses a timer based pipeline where the status is polled each time period and then the next transfer are started based on the status. 

I've now seen the callback in the DMA API but this is not used by SDMA user which is SOF. 

SOF explictly sets the callback to NULL: https://github.com/thesofproject/sof/blob/main/src/audio/dai-zephyr.c#L853

Cc: @lgirdwood

**Comment by dbaluta** [comment]
@teburd I have added the callback that will let the user know that a block transfer is finished. This is not used by the first user of the SDMA (the SOF app) but other apps can used it if they configure it.

Thanks a lot for your review!

**Comment by dbaluta** [comment]
@teburd  PR is now ready! Please have a look.

**Review by teburd — APPROVED** [review]

**Review by LaurentiuM1234 — APPROVED** [review]

**Pull request merged**